include/picongpu/plugins/openPMD/restart/LoadParticleAttributesFromOpenPMD.hpp

```diff
@@ -0,0 +1,149 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#pragma once
+
+
+#include "picongpu/plugins/openPMD/openPMDWriter.def"
+#include "picongpu/simulation_defines.hpp"
+#include "picongpu/traits/PICToOpenPMD.hpp"
+
+#include <pmacc/assert.hpp>
+#include <pmacc/traits/GetComponentsType.hpp>
+#include <pmacc/traits/GetNComponents.hpp>
+#include <pmacc/traits/Resolve.hpp>
+
+#include <openPMD/openPMD.hpp>
+
+#include <memory>
+
+namespace picongpu
+{
+namespace openPMD
+{
+    using namespace pmacc;
+
+    /** Load attribute of a species from openPMD checkpoint storage
+     *
+     * @tparam T_Identifier identifier of species attribute
+     */
+    template< typename T_Identifier >
+    struct LoadParticleAttributesFromOpenPMD
+    {
+        /** read attributes from openPMD file
+         *
+         * @param params thread params
+         * @param frame frame with all particles
+         * @param particleSpecies the openpmd representation of the species
+         * @param particlesOffset read offset in the attribute array
+         * @param elements number of elements which should be read the attribute
+         * array
+         */
+        template< typename FrameType >
+        HINLINE void
+        operator()(
+            ThreadParams * params,
+            FrameType & frame,
+            ::openPMD::ParticleSpecies particleSpecies,
+            const uint64_t particlesOffset,
+            const uint64_t elements )
+        {
+            using Identifier = T_Identifier;
+            using ValueType =
+                typename pmacc::traits::Resolve< Identifier >::type::type;
+            const uint32_t components = GetNComponents< ValueType >::value;
+            using ComponentType = typename GetComponentsType< ValueType >::type;
+            OpenPMDName< Identifier > openPMDName;
+
+            log< picLog::INPUT_OUTPUT >(
+                "openPMD: ( begin ) load species attribute: %1%" ) %
+                openPMDName();
+
+            const std::string name_lookup[] = { "x", "y", "z" };
+
+            std::shared_ptr< ComponentType > loadBfr;
+            if( elements > 0 )
+            {
+                loadBfr = std::shared_ptr< ComponentType >{
+                    new ComponentType[ elements ],
+                    []( ComponentType * ptr ) { delete[] ptr; }
+                };
+            }
+
+            for( uint32_t n = 0; n < components; ++n )
+            {
+                ::openPMD::Record record = particleSpecies[ openPMDName() ];
+                ::openPMD::RecordComponent rc = components > 1
+                    ? record[ name_lookup[ n ] ]
+                    : record[::openPMD::RecordComponent::SCALAR ];
+
+                ValueType * dataPtr =
+                    frame.getIdentifier( Identifier() ).getPointer();
+                // it's possible to aquire the local block with that call again
```

Comment:
Where does this comment belongs to? If it is a comment to the line above please move this comment up.

Reply:
That was copy-pasting from what seems to be [dead code](https://github.com/ComputationalRadiationPhysics/picongpu/blob/d340812641320d3a61e343a521287fd0720ba55b/include/picongpu/plugins/adios/restart/LoadParticleAttributesFromADIOS.hpp#L96) already in the Adios plugin. I've removed it.